internal/pkgmanager/ecosystem.go

```diff
@@ -88,7 +89,17 @@ func (p *PkgManager) DownloadArchive(name, version, directory string) (string, e
 		return "", err
 	}
 
-	archivePath, err := downloadToDirectory(directory, downloadURL)
+	fileName := ""
+	// Rename for crates (The name of downloaded archive is download)
+	if p.ecosystem == "crates.io" && path.Base(downloadURL) == "download" {
```

Comment:
Thanks for figuring out this logic! Rather than special casing here, we should split and move each case to a separate function for each ecosystem which returns the correct base filename for the URL.

Let's add a field `archiveFilename func(pkgName string, version string, downloadURL string) string` to PkgManager and implement functions for each ecosystem which have the logic you came up with below. The default value for the other ecosystems can be `path.Base(downloadURL)`.

Reply:
Should we append the names of ecosystems to the beginning of filenames?

Reply:
I don't think it's necessary. They will most likely be stored under separate directories per ecosystem in the end so it's probably fine without the ecosystem in the filename.